sdk/provisioning/Azure.Provisioning.CloudMachine/src/Tsp/TypeSpecWriter.cs

```diff
@@ -0,0 +1,244 @@
+        {
+            Type parameterType = parameter.ParameterType;
+
+            if (parameterType.FullName.Equals("Microsoft.AspNetCore.Http.HttpRequest", StringComparison.Ordinal))
+            {
+                parameterType = typeof(byte[]);
+            }
+
+            if (parameterType.IsModel())
+                models.Add(parameterType);
+
+            string location = ReadParameterLocation(parameter);
+            if (first)
+                first = false;
+            else
+                writer.Write(", ");
+
+            if (location == "@body")
+            {
+                if (parameterType == typeof(byte[]))
+                    writer.Write("@header contentType: \"application/octet-stream\", ");
+                writer.Write($"@body {parameter.Name}: {parameterType.ToTspType()}");
+            }
+            else
+                writer.Write($"{location} {parameter.Name}: {parameterType.ToTspType()}");
+        }
+        writer.WriteLine(") : {");
+        writer.WriteLine($"    @statusCode statusCode: 200;");
+        if (method.ReflectedType != typeof(void))
+            writer.WriteLine($"    @body response : {method.ReturnType.ToTspType()};");
+        writer.WriteLine("  };");
+        if (method.ReturnType.IsModel())
+            models.Add(method.ReturnType);
+    }
+
+    private static string ReadParameterLocation(ParameterInfo parameter)
+    {
+        foreach (CustomAttributeData attribute in parameter.GetCustomAttributesData())
+        {
+            if (attribute.AttributeType.Name == "FromQueryAttribute")
+                return "@query";
+            if (attribute.AttributeType.Name == "FromHeaderAttribute")
+                return "@header";
+            if (attribute.AttributeType.Name == "FromRouteAttribute")
+                return "@path";
+            if (attribute.AttributeType.Name == "FromBodyAttribute")
+                return "@body";
+        }
+        return "@body";
+    }
+    private static string ReadHttpVerb(MethodInfo method)
+    {
+        string httpVerb = "@get";
+
+        foreach (CustomAttributeData attribute in method.CustomAttributes)
+        {
+            if (attribute.AttributeType.Name == "HttpGetAttribute")
+                httpVerb = "@get";
+            else if (attribute.AttributeType.Name == "HttpPostAttribute")
+                httpVerb = "@post";
+            else if (attribute.AttributeType.Name == "HttpPutAttribute")
+                httpVerb = "@put";
+            else if (attribute.AttributeType.Name == "HttpDeleteAttribute")
+                httpVerb = "@delete";
+            else if (attribute.AttributeType.Name == "HttpPatchAttribute")
+                httpVerb = "@patch";
+            else if (attribute.AttributeType.Name == "HttpHeadAttribute")
+                httpVerb = "@head";
+        }
+
+        return httpVerb;
+    }
+
+    private static bool IsModel(this Type type)
+    {
+        return type.IsClass && !type.IsPrimitive &&
```

Comment:
Why only classes?

Reply:
we need to implement IJsonModel on them. There are serious problems with the interfaces and structs. But we need to think this whole thing through